### PR TITLE
Better native request validation

### PR DESCRIPTION
**Pull request by hhhjort**
We were requiring some fields that are only "recommended", and we have seen examples in the wild where they were not filled. This relaxes those requirements.

### CODE REVIEW

File: endpoints/openrtb2/auction.go [hunk 1]
```diff
@@ -444,6 +444,10 @@ func fillAndValidateNative(n *openrtb.Native, impIndex int) error {
 }
 
 func validateNativeContextTypes(cType native.ContextType, cSubtype native.ContextSubType, impIndex int) error {
+	if cType == 0 {
```
Comment (SyntaxNode) [inline endpoints/openrtb2/auction.go]: Can we please add a unit test or json test for this case?

File: endpoints/openrtb2/auction.go [hunk 2]
```diff
@@ -480,6 +484,10 @@ func validateNativeContextTypes(cType native.ContextType, cSubtype native.Contex
 }
 
 func validateNativePlacementType(pt native.PlacementType, impIndex int) error {
+	if pt == 0 {
```
Comment (SyntaxNode) [inline endpoints/openrtb2/auction.go]: Can we please add a unit test or json test for this case?

File: endpoints/openrtb2/auction.go (hunk 1 above)
Comment (hhhjort) [reply endpoints/openrtb2/auction.go]: endpoints/openrtb2/sample-requests/valid-native/request-no-context.json

File: endpoints/openrtb2/auction.go (hunk 2 above)
Comment (hhhjort) [reply endpoints/openrtb2/auction.go]: See the other JSON tests created/moved to reflect the new rules.

File: endpoints/openrtb2/auction.go [hunk 3]
```diff
@@ -444,6 +444,10 @@ func fillAndValidateNative(n *openrtb.Native, impIndex int) error {
 }
 
 func validateNativeContextTypes(cType native.ContextType, cSubtype native.ContextSubType, impIndex int) error {
+	if cType == 0 {
+		// Context is only recommended, so none is a valid type.
+		return nil
```
Comment (guscarreon) [inline endpoints/openrtb2/auction.go]: Should we return in here without also checking `cSubtype` first?

File: endpoints/openrtb2/auction_test.go [hunk 4]
```diff
@@ -282,7 +282,6 @@ func TestRejectAccountRequired(t *testing.T) {
 func (gr *getResponseFromDirectory) assert(t *testing.T) {
 	//t *testing.T, dir string, payloadGetter func(*testing.T, []byte) []byte, messageGetter func(*testing.T, []byte) []byte, expectedCode int, aliased bool) {
 	t.Helper()
-	var filename string
 	var filesToAssert []string
```
Comment (guscarreon) [inline endpoints/openrtb2/auction_test.go]: Great catch! `filename` was doing nothing here

File: endpoints/openrtb2/auction.go (hunk 1 above)
Comment (SyntaxNode) [reply endpoints/openrtb2/auction.go]: Ah. Changing the directory changes the validation criteria. Missed that.

File: endpoints/openrtb2/auction.go (hunk 2 above)
Comment (SyntaxNode) [reply endpoints/openrtb2/auction.go]: Ah. Changing the directory changes the validation criteria. Missed that.

File: endpoints/openrtb2/auction.go (hunk 3 above)
Comment (hhhjort) [reply endpoints/openrtb2/auction.go]: If there is no context type, would subtype have any meaning?

Comment (guscarreon) [reply endpoints/openrtb2/auction.go]: I guess not. LOL

Comment (SyntaxNode) [reply endpoints/openrtb2/auction.go]: The `contextsubtype` is optional but does have defined ranges same as cType. Not sure if we want to add it in the future, but I would think that's a separate story.

**Pull request merged**